core/sail/shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/results/ValidationResult.java

```diff
@@ -123,6 +132,58 @@ public Resource getId() {
 		return id;
 	}
 
+	private Value getAct() {
+		return Actual;
+	}
+
+	private Value getActFor() {
+		return ActualFormat;
+	}
+
+	public void SetAct(Value Ac) {
+		if (!(Ac instanceof Literal)) {
+			return;
+		}
+		this.Actual = Ac;
+		if (this.sourceConstraintComponent.equals(SourceConstraintComponent.MinCountConstraintComponent)) {
+			this.ActualFormat = Ac;
+		}
+		if (this.sourceConstraintComponent.equals(SourceConstraintComponent.DatatypeConstraintComponent)) {
+			this.ActualFormat = ((Literal) Ac).getDatatype();
+		}
+		if (this.sourceConstraintComponent.equals(SourceConstraintComponent.LanguageInConstraintComponent)) {
+			Optional<String> x = ((Literal) Ac).getLanguage();
+			if (x.isPresent()) {
+				this.ActualFormat = SimpleValueFactory.getInstance().createLiteral(x.get());
+			}
```

Comment:
This general approach will probably be hard to maintain. It would be very easy for someone to introduce a new constraint component and forget to add the required code here to handle actualFormat. I'm still not sure what the difference between actual and actualFormat is.